Add interactive elements docs (#493)
Amsterdam/design-system: storybook/storybook-docs/src/interactive-elements.stories.mdx

import { Meta } from "@storybook/blocks";

<Meta title="Docs/Designrichtlijnen/Interactieve elementen" />

> # Interactieve elementen

Reviewer: In general, please uphold the distinction between:

1. Design choices implemented in our components. These can be descriptive “Knoppen zijn blauw (omdat …)”.
2. Guidelines for using our components. These should be written in normative voice: “Zorg ervoor dat…”

Implementers need to only understand the former, but act on the latter.

You already did this in the ‘Focus indicator’ section, but it is less clear in others.

> Interactieve elementen zijn onder andere links, knoppen en formuliervelden. Interactieve elementen moeten duidelijk herkenbaar zijn en makkelijk te onderscheiden van niet-interactieve elementen. Je moet ze kunnen gebruiken op alle gangbare manieren, zoals met een muis, toetsenbord en touchscreen.

Reviewer: The start of this sentence may help define the term more explicitly:

```suggestion
Hiermee bedoelen we elementen waarmee de gebruiker kan interacteren, zoals links, knoppen en formuliervelden. Ze moeten duidelijk herkenbaar zijn en gemakkelijk te onderscheiden van niet-interactieve elementen. Je moet ze op alle gangbare manieren kunnen bedienen, zoals met een muis, toetsenbord en touchscreen.
```

## Tekst

> Interactieve elementen moeten duidelijk zijn, gebruikers moeten kunnen voorspellen wat er gebeurt als je een element activeert. Een interactief element moet ook duidelijk zijn zonder dat je de zinnen er omheen gelezen hebt. Gebruik geen onduidelijke tekst, zoals "hier" of "lees meer".

Reviewer: We can write a bit more active and terse. Separate pieces of advice can use an empty line between them. And I propose using proper curly quotes.

```suggestion
Gebruikers moeten kunnen voorspellen wat er gebeurt als je een interactief element activeert. Ook zonder de zinnen er omheen te lezen.

Gebruik geen algemene fragmenten voor links, zoals “klik hier” of “lees meer”. Dit is vervelend voor gebruikers van schermlezers die tussen links navigeren. Ook is het niet optimaal voor zoekmachines.
```

### Relevante WCAG eisen:

- [WCAG 2.4.4](https://www.w3.org/TR/WCAG22/#link-purpose-in-context): het is voor gebruikers duidelijk wat het betekent om op een link te klikken
- [WCAG 2.4.6](https://www.w3.org/TR/WCAG22/#headings-and-labels): gebruik een label om interactieve elementen te beschrijven
- [WCAG 2.5.3](https://www.w3.org/TR/WCAG22/#label-in-name): de toegankelijke naam van een element moet in ieder geval de zichtbare tekst van het element bevatten

## Kleur

> Blauw is de interactieve kleur, deze wordt bijna altijd gebruikt voor interactieve elementen. Bij hover wordt dit donkerblauw. Een uitzondering hierop zijn error states, die gebruiken rood en donkerrood voor hover.

Reviewer: ```suggestion
Om aan te geven dat een element interactief is gebruiken we de kleur blauw. Zo zijn inline links blauw en de primaire knop ook. Bij hover wordt dit donkerblauw.

Als het door andere aspecten van de visuele weergave duidelijk is dat een element interactief is, gebruiken we ook de standaard tekstkleur zwart. Een voorbeeld hiervan is een menu.

Voor foutmeldingen gebruiken we rood, en donkerrood voor hover.
```

Let op bij het inrichten van een pagina dat de interactieve elementen in een logische volgorde staan, meestal van boven naar beneden en van links naar rechts. Hierdoor krijg je een logische focus / tabbing order.

### Relevante WCAG eisen

- [WCAG 2.4.3](https://www.w3.org/TR/WCAG22/#focus-order): interactieve elementen hebben een logische focus / tabbing order
- [WCAG 2.4.7](https://www.w3.org/TR/WCAG22/#focus-visible): de focus indicator is zichtbaar
- [WCAG 2.4.11](https://www.w3.org/TR/WCAG22/#focus-not-obscured-minimum): de focus indicator wordt niet volledig verborgen door content
- [WCAG 2.4.12](https://www.w3.org/TR/WCAG22/#focus-not-obscured-enhanced): de focus indicator wordt niet deels verborgen door content
- [WCAG 2.4.13](https://www.w3.org/TR/WCAG22/#focus-appearance): de focus indicator is minimaal 2 pixels breed en heeft voldoende contrast met het element dat focus heeft

## Hover

> Gebruik verschillende, niet overlappende stijlen voor hover en focus. Als toetsenbordgebruiker is het belangrijk dat je weet welk element je activeert. Als de hover-variant (bijna) hetzelfde is als de focus-variant, dan weet je niet zeker welk element je zult activeren.

Reviewer 2: Are keyboard users the only users without a pointing device? I think we can do more for our interactive elements using `any-pointer` media queries [@media/any-pointer](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/any-pointer)

Author: > Are keyboard users the only users without a pointing device?

Keyboard users usually also have a pointing device, that's what this section is about. A user can use both tab focus and hover, which is why they should look different.

> I think we can do more for our interactive elements using any-pointer media queries @media/any-pointer

I'm not a huge fan of `any-pointer`, it's a little brittle imo. I think we should solve any issues with innacurate pointing devices in the design itself, by making thing large enough by default (which we do, currently).

We gebruiken de cursor hand `pointer` bij hover op alle interactieve elementen die geen eigen cursor stijl hebben (zoals tekstvakken). Oorspronkelijk was `pointer` alleen bedoeld voor links, maar wij vinden dat de associatie van `pointer` met interactiviteit zo sterk is dat we deze cursor vaker gebruiken.

Gebruik het `title` attribuut niet. Interactieve elementen met het `title` attribuut kunnen een tooltip tonen die over de tekst heen staat, waardoor het minder leesbaar is.

### Referenties

- [Documentatie van Utrecht over focus en hover](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/css-link--docs)
- [Specificatie van cursor pointer.](https://www.w3.org/TR/css-ui-4/#valdef-cursor-pointer) Hierin staat dat je pointer moet gebruiken voor links, en mag gebruiken voor andere interactieve elementen

> - [Artikel over het niet gebruiken van cursor pointer voor andere elementen dan links.](https://medium.com/simple-human/buttons-shouldnt-have-a-hand-cursor-b11e99ca374b)

Reviewer 2: This contradicts your statement about using pointer
> Oorspronkelijk was `pointer` alleen bedoeld voor links, maar wij vinden dat de associatie van `pointer` met interactiviteit zo sterk is dat we deze cursor vaker gebruiken.

Author: I added this to show both sides to the argument, do you think it isn't necessary?

Reviewer: I agree it may be confusing if left in without context. I can live with removing it.
